src/transport_ofi.c

```diff
@@ -84,6 +84,13 @@ uint8_t**                       shmem_transport_ofi_target_heap_addrs;
 uint8_t**                       shmem_transport_ofi_target_data_addrs;
 #endif /* ENABLE_REMOTE_VIRTUAL_ADDRESSING */
 #endif /* ENABLE_MR_SCALABLE */
+
+#ifdef USE_FI_HMEM
+struct fid_mr*                  shmem_transport_ofi_external_heap_mrfd;
+uint64_t*                       shmem_transport_ofi_external_heap_keys;
```

Comment:
Can you explain the concept of keys and addresses? Is a key essentially a password of sorts that is required to perform operations (reads/writes) on a registered memory region?

Reply:
Yes, `fi_mr_key` is the key that is used to access the memory region that is registered. After the memory registration the keys are exchanged to make sure that everyone has the remote key to access the heap address, which is also exchanged.